### PR TITLE
Fix merge of local composer template with Octane composer.json.

### PR DESCRIPTION
**Pull request by grayside**
This fixes the problem with projects intended to be used with Octane not coming through with the Octane composer.json merged in. I have annotated a few points in the diff.

### CODE REVIEW

File: generators/app/index.js [hunk 1]
```diff
@@ -174,13 +174,16 @@ module.exports = Generator.extend({
...
       if (typeof options.drupalDistro.modifyComposer == 'function') {
         var done = this.async();
-        var self = this;
-        options.drupalDistro.modifyComposer(self, options, this.composer, isNewProject, done, function(err, result, done) {
+        options.drupalDistro.modifyComposer(options, this.composer, isNewProject, done, function(err, result, done) {
           if (!err && result) {
             this.composer = result;
           }
+          else {
+            this.log.warning("Could not retrieve Octane's composer.json: "  + err);
```
Comment (grayside) [inline generators/app/index.js:182]: As an aside, I moved the logging of the error out of the distro plugin, and so was able to remove passing yo in as a parameter.

File: generators/app/index.js [hunk 2]
```diff
@@ -174,13 +174,16 @@ module.exports = Generator.extend({
...
+            return done(err);
```
Comment (grayside) [inline generators/app/index.js:183]: By returning done() here, we make sure the following catch-all done() call is not reached. Extraneous done() calls are probably safe here, but they are a bad practice in general code as extra calls can lead to prematurely terminating an asynchronous process.

File: generators/app/index.js [hunk 3]
```diff
@@ -162,39 +162,29 @@ module.exports = Generator.extend({
...
-
-      // Overwrite new project composer.json with a new core version.
-      // This needs to go down here to give distros a chance to preemptively
-      // modify the version of drupal/core.
-      if (isNewProject) {
-        // drupal/core is set, it is a real version or range and we have a
-        // release version to derive something better.
-        if (this.composer.require['drupal/core'] && !_.isString(this.composer.require['drupal/core']) && options.drupalDistroRelease) {
-          this.composer.require['drupal/core'] = drupalOrgApi.toMinorRange(options.drupalDistroRelease);
-        }
```
Comment (grayside) [inline generators/app/index.js:194]: This section is unrelated. It's some extra massaging to get a the latest drupal/core minor version number into the composer.json. I decided this was a source of problems so I'm making it go away. It just happened to get tangled with the other changes.

File: generators/app/index.js [hunk 4]
```diff
@@ -162,39 +162,29 @@ module.exports = Generator.extend({
...
+        }.bind(this));
       }
-
-      this.fs.writeJSON('composer.json', this.composer);
```
Comment (grayside) [inline generators/app/index.js:197]: The async/done combination where the composer file is modified doesn't block execution of this line. It blocks moving on to the next function or phase in the yeoman processing. As a result, the modifyComposer process wasn't blocking this. The localComposer was being written, then this.composer was assigned the new values, then the rest of the generator would execute. I moved this down below into the writing phase of the generator.

File: generators/app/index.js [hunk 5]
```diff
@@ -162,39 +162,29 @@ module.exports = Generator.extend({
...
-    composerJson: function () {
+    prepareComposerJson: function () {
       var isNewProject = (this.composerOrig == undefined);
       if (!isNewProject) {
         // Use original composer file if project already generated.
         this.composer = this.composerOrig;
       }
 
-      this.composer.name = options.projectName;
+      this.composer.name = 'organization/' + options.projectName;
```
Comment (grayside) [inline generators/app/index.js:172]: Not having a `something/` throws composer validation warnings.

**Comment by grayside** [comment]
Okay, lots of test failures. Getting into them now.

File: test/generator.int.octane.test.js [hunk 6]
```diff
@@ -39,14 +39,20 @@ describe('gadget:app for Octane', function() {
...
 
     it('does not specify a core version', function() {
       var json = JSON.parse(fs.readFileSync('composer.json', 'utf8'));
       assert.ok(!json.require['drupal/core']);
     });
+
+    it('does include merged dependencies from the Octane composer.json', function() {
+      var json = JSON.parse(fs.readFileSync('composer.json', 'utf8'));
+      assert.ok(json.require['acquia/lightning']);
+      assert.ok(json['require-dev']['se/selenium-server-standalone']);
```
Comment (grayside) [inline test/generator.int.octane.test.js:54]: Nice new test that confirms two things unique from Octane's composer.json show up in the generated composer.json.

Comment (mike-potter) [reply test/generator.int.octane.test.js:54]: Good test.  You might also add an assert that the json.require does *not* have a drupal/core entry since we want to enforce that dependency coming from lightning and not duplicated here in the project composer.json.  Some projects are claiming that yo-p2 added drupal/core along with acquia/lighting so a test to ensure this isn't happening would be good.

**Comment by mike-potter** [comment]
Other than the tweak to the new test mentioned above, I think this looks good.

File: test/generator.int.octane.test.js (hunk 6 above)
Comment (grayside) [reply test/generator.int.octane.test.js:54]: There is already a test for that a few lines up on 48, so I'll call that feedback actioned!
